### ISSUE TITLE
OTP-24.0 randomly segfault in memory allocator

### ISSUE THREAD

**Issue by hizel** [erlang/otp#4858]
**Describe the bug**
OTP-24.0 randomly segault in gc with bt, always in mbc_alloc_finalize:
```
(gdb) bt                                                                                                                                                             
#0  0x000055e91aff12a0 in mbc_alloc_finalize (allctr=0x55e91cef7540, blk=0x7fc88ee7c390, org_blk_sz=140498660610976, flags=2, crr=0x2, want_blk_sz=800,                  valid_blk_info=1) at beam/erl_alloc_util.c:2458                                                                                                                  
#1  0x000055e91aff1621 in mbc_alloc (allctr=0x55e91cef7540, size=792) at beam/erl_alloc_util.c:2530                                                                  
#2  0x000055e91affbd02 in do_erts_alcu_alloc (type=1250, allctr=0x55e91cef7540, size=792) at beam/erl_alloc_util.c:5989                                              
#3  0x000055e91affbed5 in erts_alcu_alloc_thr_spec (type=1250, extra=0x55e91b40b2f0 <erts_allctr_thr_spec+48>, size=792) at beam/erl_alloc_util.c:6063               
#4  0x000055e91b1a0595 in erts_alloc (type=1250, size=792) at beam/erl_alloc.h:251                                                                                   
#5  0x000055e91b1aaa9f in erts_hashmap_from_ks_and_vs_extra (factory=0x7fc8884e3b80, ks=0x7fc79be32350, vs=0x7fc79be32470, n=33, key=0, value=0)                         at beam/erl_map.c:711                                                                                                                                            
#6  0x000055e91afc7727 in erts_gc_update_map_assoc (p=0x7fc85c1c7be8, reg=0x7fc8884e3c80, live=0, n=33, new_p=0x7fc83d958fa8 <weak_call::codeFooter+28>)                 at beam/beam_common.c:2233                                                                                                                                       
#7  0x00007fc88c03845f in global::update_map_assoc_shared ()                                                                                                         
Backtrace stopped: previous frame inner to this frame (corrupt stack?)                                                                                               
(gdb) etp-process-info 0x7fc85c1c7be8                                                                                                                                
  Pid: <0.542.0>                                                                                                                                                       
  State: running | active | prq-prio-normal | usr-prio-normal | act-prio-normal                                                                                      
                                                                                                                                                                     
  Flags:                                                                                                                                                             
  Current function: lists:reverse/1                                                                                                                                  
  I: #Cp<lists:reverse/1+0x30>                                                                                                                                       
  Heap size: 28690                                                                                                                                                   
  Old-heap size: 46422                                                                                                                                               
  Mbuf size: 40                                                                                                                                                      
  Msgq len: 368 (inner=355, outer=13)                                                                                                                                
  Msgq Flags: on-heap                                                                                                                                                
  Parent: <0.540.0>                                                                                                                                                  
  Pointer: (Process*)0x7fc85c1c7be8                                                                        

(gdb) p/x org_blk_sz                                                                                                                                                 
$1 = 0x7fc864c1d3a0   
```

**To Reproduce**
Reproduce in my production code. I was unable to write a test.

**Affected versions**
OTP-24.0

**Additional context**
Also, with beam.debug.smp assert triggered:
```
* thread #8, name = '4_scheduler', stop reason = signal SIGABRT
    frame #0: 0x00007ffff7a6111a libc.so.6`__GI_raise(sig=<unavailable>) at raise.c:50:1
(lldb) bt
* thread #8, name = '4_scheduler', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff7a6111a libc.so.6`__GI_raise(sig=<unavailable>) at raise.c:50:1
    frame #1: 0x00007ffff7a4a53a libc.so.6`__GI_abort at abort.c:79:7
    frame #2: 0x000055555594861c beam.debug.smp`erl_assert_error(expr="rootset->size >= n", func="setup_rootset", file="beam/erl_gc.c", line=2611) at sys.c:956:5
    frame #3: 0x00005555558a37f8 beam.debug.smp`setup_rootset(p=0x00007fff8e76c548, objv=0x0000555555d70498, nobj=0, rootset=0x00007fffb41b5520) at erl_gc.c:2611:5
    frame #4: 0x000055555589f5fd beam.debug.smp`do_minor(p=0x00007fff8e76c548, live_hf_end=0xfffffffffffffff8, mature="Z\U00000010\x90\xca\xfe\U0000007f", mature_size=5976, new_sz=4185, objv=0x0000555555d70498, nobj=0) at erl_gc.c:1515:9
    frame #5: 0x000055555589eba1 beam.debug.smp`minor_collection(p=0x00007fff8e76c548, live_hf_end=0xfffffffffffffff8, need=0, objv=0x0000555555d70498, nobj=0, ygen_usage=3257, recl=0x00007fffb41b5908) at erl_gc.c:1409:9
    frame #6: 0x000055555589c023 beam.debug.smp`garbage_collect(p=0x00007fff8e76c548, live_hf_end=0xfffffffffffffff8, need=0, objv=0x0000555555d70498, nobj=0, fcalls=4000, max_young_gen_usage=0) at erl_gc.c:740:16
    frame #7: 0x000055555589ca9d beam.debug.smp`erts_garbage_collect_nobump(p=0x00007fff8e76c548, need=0, objv=0x0000555555d70498, nobj=0, fcalls=4000) at erl_gc.c:882:12
    frame #8: 0x000055555562be82 beam.debug.smp`scheduler_gc_proc(c_p=0x00007fff8e76c548, reds_left=4000) at erl_process.c:9235:12
    frame #9: 0x000055555562de29 beam.debug.smp`erts_schedule(esdp=0x00007fffb61fc980, p=0x00007fff8e76c548, calls=0) at erl_process.c:9962:32
    frame #10: 0x0000555555641ebb beam.debug.smp`::erts_debug_schedule(esdp=0x0000000000000000, c_p=0x00007fff8e76c548, calls=62) at beam_asm.cpp:323:24
```

Looks like setup_rootset from beam/erl_gc.c corrupt memory in some cases.

**Comment by hizel** [comment]
```
diff --git a/erts/emulator/beam/erl_gc.c b/erts/emulator/beam/erl_gc.c
index cb21f14071..0ffa93cf49 100644
--- a/erts/emulator/beam/erl_gc.c
+++ b/erts/emulator/beam/erl_gc.c
@@ -2579,6 +2579,7 @@ setup_rootset(Process *p, Eterm *objv, int nobj, Rootset *rootset)
 #endif
         
         len = erts_proc_sig_privqs_len(p);
+        len += 2;
 
        /* Ensure large enough rootset... */
        if (n + len > rootset->size) {
```

solve my problem, but looks bad

**Comment by rickard-green** [comment]
@hizel Great! The issue is that we do not allocate enough to take care of alias-message signal. There will soon be a patch fixing this!

**Comment by rickard-green** [comment]
@hizel I've made a PR #4870 which should fix this. It has not been thoroughly tested yet, but I'm quite confident that it wont cause any problems. I expect that it should be released in a patch early next week.

**Comment by hizel** [comment]
Thanks, this pr works for me.

**Comment by rickard-green** [comment]
The fix for this has been released in the [OTP 24.0.2](https://erlang.org/download/OTP-24.0.2.README) patch now.
